File: core/src/androidTest/java/com/emarsys/core/database/repository/AbstractSqliteRepositoryTest.java
package com.emarsys.core.database.repository;

import android.content.ContentValues;
import android.database.Cursor;
import android.support.test.InstrumentationRegistry;

import com.emarsys.core.database.CoreSQLiteDatabase;
import com.emarsys.core.database.DatabaseContract;
import com.emarsys.core.database.helper.CoreDbHelper;
import com.emarsys.core.database.helper.DbHelper;
import com.emarsys.core.database.trigger.TriggerKey;
import com.emarsys.core.provider.timestamp.TimestampProvider;
import com.emarsys.core.provider.uuid.UUIDProvider;
import com.emarsys.core.request.model.RequestModel;
import com.emarsys.testUtil.DatabaseTestUtils;
import com.emarsys.testUtil.TimeoutUtils;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TestRule;
import org.mockito.Mockito;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;

import static com.emarsys.core.database.DatabaseContract.REQUEST_COLUMN_NAME_HEADERS;
import static com.emarsys.core.database.DatabaseContract.REQUEST_COLUMN_NAME_METHOD;
import static com.emarsys.core.database.DatabaseContract.REQUEST_COLUMN_NAME_PAYLOAD;
import static com.emarsys.core.database.DatabaseContract.REQUEST_COLUMN_NAME_REQUEST_ID;
import static com.emarsys.core.database.DatabaseContract.REQUEST_COLUMN_NAME_TIMESTAMP;
import static com.emarsys.core.database.DatabaseContract.REQUEST_COLUMN_NAME_TTL;
import static com.emarsys.core.database.DatabaseContract.REQUEST_COLUMN_NAME_URL;
import static com.emarsys.core.util.serialization.SerializationUtils.serializableToBlob;
import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class AbstractSqliteRepositoryTest {

    private static class DummySqliteRepository extends AbstractSqliteRepository {

        public DummySqliteRepository(String tableName, DbHelper dbHelper) {
            super(tableName, dbHelper);
        }

        @Override
        protected ContentValues contentValuesFromItem(Object item) {
            return null;
        }

        @Override
        protected Object itemFromCursor(Cursor cursor) {
            return null;
        }

        @Override
        public void remove(Object specification) {

        }

        @Override
        public List query(Object specification) {
            return null;
        }
    }

    private final static String TABLE_NAME = "table";

    private AbstractSqliteRepository<Object> repository;
    private DbHelper dbHelperMock;
    private CoreSQLiteDatabase dbMock;

    @Rule
    public TestRule timeout = TimeoutUtils.getTimeoutRule();

    @Before
    @SuppressWarnings("unchecked")
    public void init() {
        DatabaseTestUtils.deleteCoreDatabase();

        dbMock = mock(CoreSQLiteDatabase.class);

        dbHelperMock = mock(DbHelper.class);
        when(dbHelperMock.getReadableCoreDatabase()).thenReturn(dbMock);
        when(dbHelperMock.getWritableCoreDatabase()).thenReturn(dbMock);

        repository = mock(AbstractSqliteRepository.class, Mockito.CALLS_REAL_METHODS);
        repository.tableName = TABLE_NAME;
        repository.dbHelper = dbHelperMock;
    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructor_tableName_mustNotBeNull() {
        new DummySqliteRepository(null, dbHelperMock);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testAdd_shouldNotAcceptNull() {
        repository.add(null);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testRemove_shouldNotAcceptNull() {
        repository.remove(null);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testQuery_shouldNotAcceptNull() {
        repository.query(null);
    }

    @Test
    public void testAdd_shouldInsertIntoDb() {
        ContentValues contentValues = new ContentValues();
        contentValues.put("key", "value");
        when(repository.contentValuesFromItem(any())).thenReturn(contentValues);

        Object input = new Object();

        repository.add(input);

        verify(repository).contentValuesFromItem(input);
        verify(dbMock).beginTransaction();
        verify(dbMock).insert(TABLE_NAME, null, contentValues);
        verify(dbMock).setTransactionSuccessful();
        verify(dbMock).endTransaction();
    }

    @Test
    public void testQuery_shouldReturnCorrectResult() {
        SqlSpecification specification = mock(SqlSpecification.class);
        when(specification.getSelection()).thenReturn("sql statement");
        when(specification.getSelectionArgs()).thenReturn(new String[]{"a", "b", "c"});

        Cursor cursor = mock(Cursor.class);
        when(cursor.moveToFirst()).thenReturn(true);
        when(cursor.isAfterLast()).thenReturn(false, false, false, true);

        when(dbMock.rawQuery(any(String.class), any(String[].class))).thenReturn(cursor);

        Object item1 = new Object();
        Object item2 = new Object();
        Object item3 = new Object();

        when(repository.itemFromCursor(cursor)).thenReturn(item1, item2, item3);

        List<Object> expected = Arrays.asList(item1, item2, item3);
        List<Object> result = repository.query(specification);

        verify(dbMock).rawQuery(specification.getSelection(), specification.getSelectionArgs());
        assertEquals(expected, result);
    }

    @Test
    public void testQuery_shouldReturnCorrectResult_whenCursorIsEmpty() {
        SqlSpecification specification = mock(SqlSpecification.class);
        when(specification.getSelection()).thenReturn("dummy");
        when(specification.getSelectionArgs()).thenReturn(new String[]{"value"});

        Cursor cursor = mock(Cursor.class);
        when(cursor.moveToFirst()).thenReturn(false);

        when(dbMock.rawQuery(any(String.class), any(String[].class))).thenReturn(cursor);

        List<Object> expected = Collections.emptyList();
        List<Object> result = repository.query(specification);

        assertEquals(expected, result);
    }

    @Test
    public void testRemove_shouldDeleteSpecifiedRow() {
        SqlSpecification specification = mock(SqlSpecification.class);
        when(specification.getSelection()).thenReturn("sql statement");
        when(specification.getSelectionArgs()).thenReturn(new String[]{"a", "b", "c"});

        repository.remove(specification);

        verify(dbMock).beginTransaction();
        verify(dbMock).delete(TABLE_NAME, specification.getSelection(), specification.getSelectionArgs());
        verify(dbMock).setTransactionSuccessful();
        verify(dbMock).endTransaction();
    }

    @Test
    public void testIsEmpty_shouldReturnFalse_whenThereAreRows() {
        DbHelper helper = new CoreDbHelper(
                InstrumentationRegistry.getTargetContext(),
                new HashMap<TriggerKey, List<Runnable>>());
        repository.tableName = DatabaseContract.REQUEST_TABLE_NAME;
        repository.dbHelper = helper;
        CoreSQLiteDatabase db = helper.getWritableCoreDatabase();
        TimestampProvider timestampProvider = new TimestampProvider();
        UUIDProvider uuidProvider = new UUIDProvider();

        RequestModel model1 = new RequestModel.Builder(timestampProvider, uuidProvider).url("https://google.com").build();
        RequestModel model2 = new RequestModel.Builder(timestampProvider, uuidProvider).url("https://emarsys.com").build();
        db.insert(DatabaseContract.REQUEST_TABLE_NAME, null, contentValuesFrom(model1));
        db.insert(DatabaseContract.REQUEST_TABLE_NAME, null, contentValuesFrom(model2));

        assertFalse(repository.isEmpty());
    }

    @Test
    public void testIsEmpty_shouldReturnTrue_whenTableIsEmpty() {
        DbHelper helper = new CoreDbHelper(
                InstrumentationRegistry.getTargetContext(),
                new HashMap<TriggerKey, List<Runnable>>());
        repository.dbHelper = helper;
        repository.tableName = DatabaseContract.REQUEST_TABLE_NAME;
        CoreSQLiteDatabase db = helper.getWritableCoreDatabase();

        db.execSQL(String.format("DELETE FROM %s;", DatabaseContract.REQUEST_TABLE_NAME));

        assertTrue(repository.isEmpty());
    }

    private ContentValues contentValuesFrom(RequestModel item) {
        ContentValues contentValues = new ContentValues();
        contentValues.put(REQUEST_COLUMN_NAME_REQUEST_ID, item.getId());
        contentValues.put(REQUEST_COLUMN_NAME_METHOD, item.getMethod().name());
        contentValues.put(REQUEST_COLUMN_NAME_URL, item.getUrl().toString());
        contentValues.put(REQUEST_COLUMN_NAME_HEADERS, serializableToBlob(item.getHeaders()));
        contentValues.put(REQUEST_COLUMN_NAME_PAYLOAD, serializableToBlob(item.getPayload()));
        contentValues.put(REQUEST_COLUMN_NAME_TIMESTAMP, item.getTimestamp());
        contentValues.put(REQUEST_COLUMN_NAME_TTL, item.getTtl());
        return contentValues;
    }
}

File: core/src/androidTest/java/com/emarsys/core/database/repository/AbstractSqliteRepositoryTest.kt
package com.emarsys.core.database.repository

import android.content.ContentValues
import android.database.Cursor
import com.emarsys.core.concurrency.ConcurrentHandlerHolderFactory
import com.emarsys.core.database.CoreSQLiteDatabase
import com.emarsys.core.database.DatabaseContract
import com.emarsys.core.database.DatabaseContract.REQUEST_COLUMN_NAME_HEADERS
import com.emarsys.core.database.DatabaseContract.REQUEST_COLUMN_NAME_METHOD
import com.emarsys.core.database.DatabaseContract.REQUEST_COLUMN_NAME_PAYLOAD
import com.emarsys.core.database.DatabaseContract.REQUEST_COLUMN_NAME_REQUEST_ID
import com.emarsys.core.database.DatabaseContract.REQUEST_COLUMN_NAME_TIMESTAMP
import com.emarsys.core.database.DatabaseContract.REQUEST_COLUMN_NAME_TTL
import com.emarsys.core.database.DatabaseContract.REQUEST_COLUMN_NAME_URL
import com.emarsys.core.database.helper.CoreDbHelper
import com.emarsys.core.database.helper.DbHelper
import com.emarsys.core.device.FilterByClientId
import com.emarsys.core.handler.ConcurrentHandlerHolder
import com.emarsys.core.provider.timestamp.TimestampProvider
import com.emarsys.core.provider.uuid.UUIDProvider
import com.emarsys.core.request.model.RequestModel
import com.emarsys.core.util.serialization.SerializationUtils.serializableToBlob
import com.emarsys.testUtil.DatabaseTestUtils
import com.emarsys.testUtil.InstrumentationRegistry
import com.emarsys.testUtil.mockito.anyNotNull
import io.kotest.matchers.shouldBe
import kotlinx.coroutines.runBlocking
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito
import org.mockito.kotlin.any
import org.mockito.kotlin.anyOrNull
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.doThrow
import org.mockito.kotlin.eq
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever

class AbstractSqliteRepositoryTest  {

    companion object {
        private const val DISTINCT = true
        private const val TABLE_NAME = "table"
        private val COLUMNS = arrayOf("col1", "col2", "col3")
        private const val SELECTION = "selection"
        private val SELECTION_ARGS = arrayOf("arg1")
        private const val GROUP_BY = "group by"
        private const val HAVING = "having condition"
        private const val ORDER_BY = "order by"
        private const val LIMIT = "limit amount"
    }

    private lateinit var repository: AbstractSqliteRepository<Any>
    private lateinit var dbHelperMock: DbHelper
    private lateinit var dbMock: CoreSQLiteDatabase
    private lateinit var dummySpecification: SqlSpecification
    private lateinit var testConcurrentHandlerHolder: ConcurrentHandlerHolder


    @Before
    @Suppress("UNCHECKED_CAST")
    fun init() {
        DatabaseTestUtils.deleteCoreDatabase()
        testConcurrentHandlerHolder =
            ConcurrentHandlerHolderFactory.create()

        dummySpecification = sqlSpecification(
            DISTINCT,
            COLUMNS,
            SELECTION,
                SELECTION_ARGS,
                GROUP_BY,
                HAVING,
                ORDER_BY,
                LIMIT
        )

        dbMock = mock()

        dbHelperMock = mock {
            on { readableCoreDatabase } doReturn dbMock
            on { writableCoreDatabase } doReturn dbMock
        }

        repository = (mock(defaultAnswer = Mockito.CALLS_REAL_METHODS) as AbstractSqliteRepository<Any>).apply {
            tableName = TABLE_NAME
            dbHelper = dbHelperMock
            concurrentHandlerHolder = testConcurrentHandlerHolder
        }
    }

    @Test
    fun testAdd_shouldInsertIntoDb() {
        val contentValues = ContentValues().apply {
            put("key", "value")
        }
        whenever(repository.contentValuesFromItem(anyOrNull())).thenReturn(contentValues)

        val input = any<Any>()

        runBlocking {
            repository.add(input)
        }

        verify(repository).contentValuesFromItem(input)
        verify(dbMock).beginTransaction()
        verify(dbMock).insert(TABLE_NAME, null, contentValues)
        verify(dbMock).setTransactionSuccessful()
        verify(dbMock).endTransaction()
    }

    @Test
    fun testUpdate_shouldUpdateTheDb() {
        runBlocking {
            val contentValues = ContentValues().apply {
                put("key", "value")
            }
            whenever(repository.contentValuesFromItem(anyNotNull())).thenReturn(contentValues)
            whenever(dbMock.update(any(), any(), any(), any())).thenReturn(1)
            val input = anyNotNull<Any>()

            val result = repository.update(input, eq(FilterByClientId("id")))

            verify(repository).contentValuesFromItem(input)
            verify(dbMock).beginTransaction()
            verify(dbMock).update(TABLE_NAME, contentValues, "hardware_id=?", arrayOf("id"))
            verify(dbMock).setTransactionSuccessful()
            verify(dbMock).endTransaction()
            result shouldBe 1
        }
    }

    @Test
    fun testQuery_shouldReturnCorrectResult() {
        val cursor = mock<Cursor> {
            on { moveToFirst() } doReturn true
        }
        whenever(cursor.isAfterLast).thenReturn(false, false, false, true)
        whenever(dbMock.query(
                DISTINCT,
                TABLE_NAME,
                COLUMNS,
                SELECTION,
                SELECTION_ARGS,
                GROUP_BY,
                HAVING,
                ORDER_BY,
                LIMIT)).thenReturn(cursor)

        val item1 = Any()
        val item2 = Any()
        val item3 = Any()

        whenever(repository.itemFromCursor(cursor)).thenReturn(item1, item2, item3)

        val result = repository.query(dummySpecification)

        verify(dbMock).query(
                DISTINCT,
                TABLE_NAME,
                COLUMNS,
                SELECTION,
                SELECTION_ARGS,
                GROUP_BY,
                HAVING,
                ORDER_BY,
                LIMIT)

        result shouldBe arrayOf(item1, item2, item3)
    }

    @Test
    fun testQuery_shouldReturnCorrectResult_whenCursorIsEmpty() {
        val cursor = mock<Cursor> {
            on { moveToFirst() } doReturn false
        }

        whenever(dbMock.query(
                DISTINCT,
                TABLE_NAME,
                COLUMNS,
                SELECTION,
                SELECTION_ARGS,
                GROUP_BY,
                HAVING,
                ORDER_BY,
                LIMIT)).thenReturn(cursor)

        val result = repository.query(dummySpecification)

        result shouldBe emptyList()
    }

    @Test
    fun testRemove_shouldDeleteSpecifiedRow() {

        runBlocking {
            repository.remove(dummySpecification)
        }

        verify(dbMock).beginTransaction()
        verify(dbMock).delete(
            TABLE_NAME,
            dummySpecification.selection,
            dummySpecification.selectionArgs
        )
        verify(dbMock).setTransactionSuccessful()
        verify(dbMock).endTransaction()
    }

    @Test
    fun testIsEmpty_shouldReturnFalse_whenThereAreRows() {
        val helper = CoreDbHelper(
                InstrumentationRegistry.getTargetContext(),
                HashMap())

        repository.dbHelper = helper
        repository.tableName = DatabaseContract.REQUEST_TABLE_NAME
        val db = helper.writableCoreDatabase

        listOf("https://google.com", "https://emarsys.com")
                .map(this::requestModel)
                .forEach {
                    db.insert(DatabaseContract.REQUEST_TABLE_NAME, null, contentValuesFrom(it))
                }

        repository.isEmpty() shouldBe false
    }

    @Test
    fun testIsEmpty_shouldReturnTrue_whenTableIsEmpty() {
        val helper = CoreDbHelper(
                InstrumentationRegistry.getTargetContext(),
                HashMap())

        repository.dbHelper = helper
        repository.tableName = DatabaseContract.REQUEST_TABLE_NAME
        val db = helper.writableCoreDatabase

        db.execSQL("DELETE FROM request;")

        repository.isEmpty() shouldBe true
    }

    @Test
    fun testCursor_shouldBeClosed_afterException() {
        val mockCursor = mock<Cursor> {
            on { moveToFirst() } doThrow RuntimeException("TestException")
        }
        whenever(dbMock.query(any(), any(), any(), any(), any(), any(), any(), any(), any())).thenReturn(mockCursor)
        try {
            repository.query(dummySpecification)
        } catch (ignored: RuntimeException) {
        }

        verify(mockCursor).close()
    }

    private fun contentValuesFrom(item: RequestModel): ContentValues {
        val contentValues = ContentValues()
        contentValues.put(REQUEST_COLUMN_NAME_REQUEST_ID, item.id)
        contentValues.put(REQUEST_COLUMN_NAME_METHOD, item.method.name)
        contentValues.put(REQUEST_COLUMN_NAME_URL, item.url.toString())
        contentValues.put(REQUEST_COLUMN_NAME_HEADERS, serializableToBlob(item.headers))
        contentValues.put(REQUEST_COLUMN_NAME_PAYLOAD, serializableToBlob(item.payload))
        contentValues.put(REQUEST_COLUMN_NAME_TIMESTAMP, item.timestamp)
        contentValues.put(REQUEST_COLUMN_NAME_TTL, item.ttl)
        return contentValues
    }

    private fun sqlSpecification(
            distinct: Boolean,
            columns: Array<String>,
            selection: String,
            selectionArgs: Array<String>,
            groupBy: String,
            having: String,
            orderBy: String,
            limit: String) = object : SqlSpecification {

        override val isDistinct: Boolean
            get() = distinct
        override val columns: Array<String>
            get() = columns
        override val selection: String
            get() = selection
        override val selectionArgs: Array<String>
            get() = selectionArgs
        override val groupBy: String
            get() = groupBy
        override val having: String
            get() = having
        override val orderBy: String
            get() = orderBy
        override val limit: String
            get() = limit
    }

    private fun requestModel(url: String) = RequestModel.Builder(TimestampProvider(), UUIDProvider()).url(url).build()
}